torch/testing/_internal/composite_compliance.py

```diff
@@ -0,0 +1,174 @@
+import torch
+import contextlib
+from typing import Iterator
+from torch.utils._pytree import tree_map
+from functools import partial
+from torch.utils._python_dispatch import enable_python_mode
+
+# TODO: move this into library proper
+@contextlib.contextmanager
+def no_dispatch() -> Iterator[None]:
+    guard = torch._C._DisableTorchDispatch()  # type: ignore[attr-defined]
+    try:
+        yield
+    finally:
+        del guard
+
+def check_metadata_consistency(wrapper_tensor):
+    if not isinstance(wrapper_tensor, CompositeCompliantTensor):
+        return
+    elem = wrapper_tensor.elem
+    if wrapper_tensor.shape != elem.shape:
+        raise RuntimeError(
+            "This operator is not CompositeImplicitAutograd compliant: the "
+            "shape of the tensor was modified directly without "
+            "going through the PyTorch dispatcher.")
+    if wrapper_tensor.dtype != elem.dtype:
+        raise RuntimeError(
+            "This operator is not CompositeImplicitAutograd compliant: the "
+            "dtype of the tensor was modified directly without "
+            "going through the PyTorch dispatcher.")
+
+# This is a bit fragile because this needs to be updated whenever a new view op
+# is added. We probably want some sort of introspection to tell that if a
+# torch.* op is a view or not.
+def is_view_fn(func):
+    return func.__name__ in {
+        'as_strided',
+        'detach',
+        'diagonal',
+        'expand',
+        'expand_as',
+        'movedim',
+        'narrow',
+        'permute',
+        'select',
+        'squeeze',
+        'transpose',
+        't',
+        'real',
+        'imag',
+        'view_as_real',
+        'view_as_complex',
+        'unflatten',
+        'unfold',
+        'unsqueeze',
+        'view',
+        'view_as',
+        'unbind',
+        'split',
+        'split_with_sizes',
+        'vsplit',
+        'hsplit',
+        'tensor_split',
+        'chunk',
+        'swapaxes',
+        'slice',
+        '_reshape_alias',
+        '_unsafe_view',
+        '_conj',
+    }
+
+def is_inplace_view_fn(func):
```

Comment:
assuming that you can do away with the hardcoded view list and just need to deal with view_inplace ops: I need this list for functionalization too, and I'm doing it through tagging in native_functions.yaml. At some point that'll probably be available for introspection (probably not before this PR lands though).